Zend/zend_objects.c

```diff
@@ -279,6 +279,52 @@ ZEND_API void ZEND_FASTCALL zend_objects_clone_members(zend_object *new_object,
 	}
 }
 
+ZEND_API zend_object *zend_objects_clone_obj_with(zend_object *old_object, const zend_class_entry *scope, const HashTable *properties)
+{
+	zend_object *new_object;
+
+	new_object = old_object->handlers->clone_obj(old_object);
+
+	if (EXPECTED(!EG(exception)) && zend_hash_num_elements(properties) > 0) {
+		GC_ADDREF(new_object);
+
+		/* Unlock readonly properties once more. */
+		if (ZEND_CLASS_HAS_READONLY_PROPS(new_object->ce)) {
+			for (uint32_t i = 0; i < new_object->ce->default_properties_count; i++) {
+				zval* prop = OBJ_PROP_NUM(new_object, i);
+				Z_PROP_FLAG_P(prop) |= IS_PROP_REINITABLE;
+			}
+		}
+
+		const zend_class_entry *old_scope = EG(fake_scope);
+
+		EG(fake_scope) = scope;
+
+		zend_ulong num_key;
+		zend_string *key;
+		zval *val;
+		ZEND_HASH_FOREACH_KEY_VAL(properties, num_key, key, val) {
```

Comment:
The following code will break:
```php
$lol = 'I love references';
$x = new stdClass;
$y = clone($x, ['x' => &$lol]);
var_dump($y);
```

Reply:
Good catch. Behavior with references is not clearly specified in the RFC, but following the showcased “desugaring” of:

```
$cloned = clone $object;
foreach ($withProperties as $key => $value) {
    $cloned->{$key} = $value;
}
return $cloned;
```

would mean that unwrapping references would be the correct solution. Alternatively, we could also throw an `Error` that references are unsupported and defer the question to the future. What do you think @edorian?

Reply:
We decided to make references an error for now and will follow-up with another RFC if someone complains / if the need arises.

Reply:
FWIW, I also think unwrapping the references is the correct solution, as this is what happens for normal assignments. You can easily encounter references even if they are unique, e.g. by looping over the array that is passed to `clone`.

Reply:
> I think it would be reasonable to at least unwrap references if they have RC=1, as these are most-likely accidental.

Okay, will look into that tomorrow.

Reply:
Updated. As this should've been everything, I've also squashed my fixup commits and rebased onto the latest master to make sure CI tests it properly.

Please take another look (and hopefully approve) :smiley: